### PR TITLE
Fixed #26424 -- Allowed bypassing schema validation of URLValidator.

### PR DESCRIPTION
**Pull request by burhan**
https://code.djangoproject.com/ticket/26424

### CODE REVIEW

File: django/core/validators.py [hunk 1]
```diff
@@ -100,11 +102,105 @@ class URLValidator(RegexValidator):
         r'^(?:[a-z0-9\.\-\+]*)://'  # scheme is validated separately
         r'(?:\S+(?::\S*)?@)?'  # user:pass authentication
         r'(?:' + ipv4_re + '|' + ipv6_re + '|' + host_re + ')'
-        r'(?::\d{2,5})?'  # port
-        r'(?:[/?#][^\s]*)?'  # resource path
-        r'\Z', re.IGNORECASE)
+                                                           r'(?::\d{2,5})?'  # port
+                                                           r'(?:[/?#][^\s]*)?'  # resource path
+                                                           r'\Z', re.IGNORECASE)
```
Comment (Dunedan) [inline django/core/validators.py]: Are these style changes intentional? Somehow they don't look like that.

Comment (burhan) [inline django/core/validators.py]: I am not sure how these got there, I blame the pep8 auto-formatter; fixed.

File: django/core/validators.py [hunk 2]
```diff
@@ -87,12 +89,13 @@ class URLValidator(RegexValidator):
...
     domain_re = r'(?:\.(?!-)[a-z' + ul + r'0-9-]{1,63}(?<!-))*'
     tld_re = (
-        '\.'                                # dot
-        '(?!-)'                             # can't start with a dash
-        '(?:[a-z' + ul + '-]{2,63}'         # domain label
-        '|xn--[a-z0-9]{1,59})'              # or punycode label
-        '(?<!-)'                            # can't end with a dash
-        '\.?'                               # may have a trailing dot
+
+        '\.'  # dot
```
Comment (timgraham) [inline django/core/validators.py]: Please remove the unrelated style changes.

File: django/core/validators.py [hunk 3]
```diff
@@ -104,7 +107,101 @@ class URLValidator(RegexValidator):
         r'(?:[/?#][^\s]*)?'  # resource path
         r'\Z', re.IGNORECASE)
     message = _('Enter a valid URL.')
-    schemes = ['http', 'https', 'ftp', 'ftps']
+    schemes = [
```
Comment (timgraham) [inline django/core/validators.py]: Changing this list is outside the scope of this ticket and also backwards-incompatible.

Comment (burhan) [inline django/core/validators.py]: Can you explain how this is backwards incompatible? It contains all the valid schemas, including the ones already included in the previous version.

Comment (timgraham) [inline django/core/validators.py]: It allows many new schemes that weren't allowed before.

Comment (burhan) [inline django/core/validators.py]: I see. So you would like it to reject still valid schemes (such as `file`, `git`), but add the ability to ignore the validation completely?

Comment (timgraham) [inline django/core/validators.py]: Yes, that's the idea. I'm not sure if we need to include a long list of schemes though. As Claude suggested on the ticket, some minimal validation like alphanumeric + period is probably sufficient. That way we don't have to maintain a list. Any projects that require such a list can supply it.

**Comment by timgraham** [comment]
Please [rebase your branch](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/working-with-git/#rebasing-branches) so that the unrelated commits don't appear.

File: django/core/validators.py (hunk 2 above)
Comment (timgraham) [inline django/core/validators.py]: There are still many unrelated formatting changes that should be reverted.

File: docs/ref/validators.txt [hunk 4]
```diff
@@ -161,9 +161,11 @@ to, or in lieu of custom ``field.clean()`` methods.
 
     .. attribute:: schemes
 
-        URL/URI scheme list to validate against. If not provided, the default
-        list is ``['http', 'https', 'ftp', 'ftps']``. As a reference, the IANA
-        website provides a full list of `valid URI schemes`_.
+        URL/URI scheme list to validate against. If not provided, the
+        default is ``['http', 'https', 'ftp', 'ftps']``. If the special
+        valid ``__all__`` is passed, then all schemas are allowed.
```
Comment (timgraham) [inline docs/ref/validators.txt]: Need a `.. versionchanged:: 1.10` annotation and a mention in the release notes. See the [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist).

File: django/core/validators.py [hunk 5]
```diff
@@ -113,10 +115,13 @@ def __init__(self, schemes=None, **kwargs):
 
     def __call__(self, value):
         value = force_text(value)
-        # Check first if the scheme is valid
-        scheme = value.split('://')[0].lower()
-        if scheme not in self.schemes:
-            raise ValidationError(self.message, code=self.code)
+        # If scheme is of special value __all__, ignore the scheme
+        # checks.
+        if self.schemes != '__all__':
```
Comment (timgraham) [inline django/core/validators.py]: As Claude suggested on the ticket, instead of skipping validation entirely, some minimal validation like alphanumeric + period could be nice. The documentation should describe the validation.

File: docs/ref/validators.txt [hunk 6]
```diff
@@ -161,14 +161,22 @@ to, or in lieu of custom ``field.clean()`` methods.
 
     .. attribute:: schemes
 
+        .. versionchanged:: 1.10
```
Comment (timgraham) [inline docs/ref/validators.txt]: This should appear at the bottom of the section with a description of what changed.

File: docs/ref/validators.txt [hunk 7]
```diff
@@ -161,14 +161,22 @@ to, or in lieu of custom ``field.clean()`` methods.
...
+
         URL/URI scheme list to validate against. If not provided, the
-        default is ``['http', 'https', 'ftp', 'ftps']``. If the special
-        valid ``__all__`` is passed, then all schemas are allowed.
+        default is ``['http', 'https', 'ftp', 'ftps']``.
+
+        Schemes must start with a letter, and may optionally contain
+        an addition protocol part appended with a `+`, ie. `git+ssh`.
```
Comment (timgraham) [inline docs/ref/validators.txt]: ie -> e.g.
must double the backticks for them to have an effect on the rendered docs

File: docs/ref/validators.txt [hunk 8]
```diff
@@ -161,14 +161,22 @@ to, or in lieu of custom ``field.clean()`` methods.
...
     .. attribute:: schemes
 
+        .. versionchanged:: 1.10
+
         URL/URI scheme list to validate against. If not provided, the
-        default is ``['http', 'https', 'ftp', 'ftps']``. If the special
-        valid ``__all__`` is passed, then all schemas are allowed.
+        default is ``['http', 'https', 'ftp', 'ftps']``.
+
+        Schemes must start with a letter, and may optionally contain
```
Comment (timgraham) [inline docs/ref/validators.txt]: no comma since "may optionally contain...." can't be a sentence on its own.

File: docs/ref/validators.txt [hunk 9]
```diff
@@ -161,14 +161,22 @@ to, or in lieu of custom ``field.clean()`` methods.
...
+        an addition protocol part appended with a `+`, ie. `git+ssh`.
+
+        If the special value ``__all__`` is passed, schemes are not
```
Comment (timgraham) [inline docs/ref/validators.txt]: '__all__' (add apostrophes)

**Comment by timgraham** [comment]
Please squash commits and use the commit message from the PR title. Also uncheck "Patch needs improvement" on the ticket so that it appears in the review queue.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: django/core/validators.py [hunk 10]
```diff
@@ -17,13 +17,15 @@
 
 def _lazy_re_compile(regex, flags=0):
     """Lazily compile a regex with flags."""
+
```
Comment (timgraham) [inline django/core/validators.py]: Please revert.

File: django/core/validators.py [hunk 11]
```diff
@@ -17,13 +17,15 @@
...
     def _compile():
         # Compile the regex if it was not passed pre-compiled.
         if isinstance(regex, six.string_types):
             return re.compile(regex, flags)
         else:
             assert not flags, "flags must be empty if regex is passed pre-compiled"
             return regex
+
```
Comment (timgraham) [inline django/core/validators.py]: Please revert.

File: django/core/validators.py (hunk 2 above)
Comment (timgraham) [inline django/core/validators.py]: Please revert all the comment movements.

File: django/core/validators.py [hunk 12]
```diff
@@ -151,6 +164,7 @@ def __call__(self, value):
         if len(urlsplit(value).netloc) > 253:
             raise ValidationError(self.message, code=self.code)
 
+
```
Comment (timgraham) [inline django/core/validators.py]: Please revert.

File: django/core/validators.py [hunk 13]
```diff
@@ -232,6 +246,7 @@ def __eq__(self, other):
             (self.code == other.code)
         )
 
+
```
Comment (timgraham) [inline django/core/validators.py]: Please revert.

File: django/core/validators.py [hunk 14]
```diff
@@ -266,6 +281,7 @@ def validate_ipv46_address(value):
         except ValidationError:
             raise ValidationError(_('Enter a valid IPv4 or IPv6 address.'), code='invalid')
 
+
```
Comment (timgraham) [inline django/core/validators.py]: Please revert.

File: django/core/validators.py [hunk 15]
```diff
@@ -322,9 +338,9 @@ def __call__(self, value):
     def __eq__(self, other):
         return (
             isinstance(other, self.__class__) and
-            self.limit_value == other.limit_value and
-            self.message == other.message and
-            self.code == other.code
+            (self.limit_value == other.limit_value) and
```
Comment (timgraham) [inline django/core/validators.py]: Please revert.

File: django/core/validators.py [hunk 16]
```diff
@@ -435,13 +451,13 @@ def __call__(self, value):
...
                 params={'max': self.decimal_places},
             )
-        if (self.max_digits is not None and self.decimal_places is not None and
-                whole_digits > (self.max_digits - self.decimal_places)):
-            raise ValidationError(
-                self.messages['max_whole_digits'],
-                code='max_whole_digits',
-                params={'max': (self.max_digits - self.decimal_places)},
-            )
+        if (self.max_digits is not None and
```
Comment (timgraham) [inline django/core/validators.py]: Please revert.

File: tests/validators/tests.py [hunk 17]
```diff
@@ -219,6 +219,7 @@
     # Trailing junk does not take forever to reject
     (URLValidator(), 'http://www.asdasdasdasdsadfm.com.br ', ValidationError),
     (URLValidator(), 'http://www.asdasdasdasdsadfm.com.br z', ValidationError),
+    (URLValidator(schemes='__all__'), 'foo://hello.com', None),
```
Comment (timgraham) [inline tests/validators/tests.py]: How about some tests for `schemes='__all__'` and URLs with schemes that don't match the regular expression?

**Comment by timgraham** [comment]
Please include a mention in the 1.10 release notes.

File: django/core/validators.py [hunk 18]
```diff
@@ -113,11 +115,24 @@ def __init__(self, schemes=None, **kwargs):
...
-        # Check first if the scheme is valid
         scheme = value.split('://')[0].lower()
-        if scheme not in self.schemes:
+
+        # A naive check for valid schemes
+        # schemes may contain an optional +, but only if followed
+        # by one or more characters
+        # RFC 3986 has the complete details
+        # http://tools.ietf.org/html/rfc3986#section-3.1
+        scheme_regex = _lazy_re_compile(r'^[a-z][a-z0-9]+(\+[a-z0-9]+)?$')
```
Comment (timgraham) [inline django/core/validators.py]: Actually the idea is to do this validation only if `schemes == '__all__'`. If the user provides some other scheme, then we might as well allow it.

Comment (burhan) [inline django/core/validators.py]: I'm a bit confused here; so either we do some basic checks no matter what schemes we are validating - this would mean moving the regular expression check higher in the queue; but then if the magic value `__all__` is passed, we accept all schemes even if they are not matching the regular expression.

Did I understand it correctly? I thought previously it was that even if `__all__` is passed, we should still do some sanity checks to warn about any "obvious" incorrect schemes.

Comment (timgraham) [inline django/core/validators.py]: Only do the regular expression check if `schemes == '__all__'`.

**Comment by timgraham** [comment]
This looks mostly good but can you replace remove the unrelated whitespace changes?

**Comment by burhan** [comment]
Hi Tim, I did not make any whitespace changes (just did a rebase from master). Can you tell me which part?

**Comment by timgraham** [comment]
Click on the "Files changed" tab of this PR and you should see them.

File: django/core/validators.py [hunk 19]
```diff
@@ -56,8 +56,7 @@ def __call__(self, value):
         Validates that the input matches the regular expression
         if inverse_match is False, otherwise raises ValidationError.
         """
-        if not (self.inverse_match is not bool(self.regex.search(
-                force_text(value)))):
+        if not (self.inverse_match is not bool(self.regex.search(force_text(value)))):
```
Comment (timgraham) [inline django/core/validators.py]: Please remove this unrelated change.

File: django/core/validators.py [hunk 20]
```diff
@@ -87,12 +86,12 @@ class URLValidator(RegexValidator):
     # Max length for domain name labels is 63 characters per RFC 1034 sec. 3.1
     domain_re = r'(?:\.(?!-)[a-z' + ul + r'0-9-]{1,63}(?<!-))*'
     tld_re = (
-        '\.'                                # dot
-        '(?!-)'                             # can't start with a dash
-        '(?:[a-z' + ul + '-]{2,63}'         # domain label
-        '|xn--[a-z0-9]{1,59})'              # or punycode label
-        '(?<!-)'                            # can't end with a dash
-        '\.?'                               # may have a trailing dot
+        '\.'  # dot
```
Comment (timgraham) [inline django/core/validators.py]: Please remove this unrelated change.

File: django/core/validators.py [hunk 21]
```diff
@@ -435,13 +447,13 @@ def __call__(self, value):
...
                 params={'max': self.decimal_places},
             )
-        if (self.max_digits is not None and self.decimal_places is not None and
-                whole_digits > (self.max_digits - self.decimal_places)):
-            raise ValidationError(
-                self.messages['max_whole_digits'],
-                code='max_whole_digits',
-                params={'max': (self.max_digits - self.decimal_places)},
-            )
+        if (self.max_digits is not None and
```
Comment (timgraham) [inline django/core/validators.py]: Please remove this unrelated change.

**Comment by timgraham** [comment]
I made some edits you can apply from this diff: http://dpaste.com/2SDVK38
However, I noticed that the RFC says schemes should also accept dashes and periods so I think you need to make that improvement as well.

**Comment by burhan** [comment]
Hey @timgraham anything pending on this one?

**Comment by timgraham** [comment]
Don't forget to uncheck "Patch needs improvement" on the ticket so it returns to the review queue. Anyway, I've been busy getting 1.10 alpha out the door and haven't had a lot of review time this week. Will get to it in due time.

**Comment by timgraham** [comment]
I wrote [a test script](https://github.com/timgraham/bug-files/tree/master/django-26424) and there are still some valid schemes from the IANA list such as ms-transit-to that aren't passing the current regex.

Also the patch needs to be targeted for 1.11 now.

**Comment by timgraham** [comment]
Closing due to inactivity. Please send a new PR if you update it.
